File: sdk/go/ota/ota.go
// Package ota provides client-side firmware OTA orchestration. It computes
// SHA-256 of the firmware blob, calls Device.OTABegin, streams Device.OTAChunk
// in sequential offsets, and finishes with Device.OTAFinalize. The device
// reboots into the new image after Finalize ACKs — the next round-trip will
// fail until the caller reconnects.
//
// Only useful when the firmware was compiled with -DCONDUYT_OTA. Check
// device.Capabilities().OTACapable before starting; the device will NAK
// OTA_BEGIN with OTA_INVALID (0x0F) if OTA isn't built in.
package ota

import (
	"context"
	"crypto/sha256"
	"errors"

	conduyt "github.com/virgilvox/conduyt/sdk/go"
)

// Options tunes the flash run.
type Options struct {
	// ChunkSize is bytes per OTA_CHUNK packet. Defaults to (max_payload - 4)
	// from HELLO_RESP, or 240 if the device hasn't advertised one.
	ChunkSize int
	// OnProgress fires after every ACKed chunk. (sent, total) in bytes.
	OnProgress func(sent, total int)
	// SHA256 is an optional pre-computed digest. If nil, computed via
	// crypto/sha256 from the firmware bytes.
	SHA256 []byte
}

// Flash uploads `firmware` to `device` and triggers a reboot into the new
// image. Returns when OTA_FINALIZE acks; the device disappears shortly after.
func Flash(ctx context.Context, device *conduyt.Device, firmware []byte, opts Options) error {
	total := len(firmware)
	if total == 0 {
		return errors.New("OTA: firmware is empty")
	}

	digest := opts.SHA256
	if digest == nil {
		sum := sha256.Sum256(firmware)
		digest = sum[:]
	}

	// Pick a chunk size. Reserve 4 bytes for the LE u32 offset header.
	size := opts.ChunkSize
	if size <= 0 {
		safeMax := 240
		if caps := device.Capabilities(); caps != nil && caps.MaxPayload > 0 {
			s := int(caps.MaxPayload) - 4
			if s > safeMax {
				safeMax = s
			} else if s < 64 {
				safeMax = 64
			} else {
				safeMax = s
			}
		}
		size = safeMax
	}

	if err := device.OTABegin(ctx, uint32(total), digest); err != nil {
		return err
	}

	sent := 0
	for sent < total {
		end := sent + size
		if end > total {
			end = total
		}
		if err := device.OTAChunk(ctx, uint32(sent), firmware[sent:end]); err != nil {
			return err
		}
		sent = end
		if opts.OnProgress != nil {
			opts.OnProgress(sent, total)
		}
	}

	return device.OTAFinalize(ctx)
}

File: sdk/python/src/conduyt/ota.py
"""CONDUYT OTA — Over-the-Air firmware update orchestrator.

Wraps the device.ota_begin / ota_chunk / ota_finalize primitives with:
  - SHA-256 of the firmware image computed via hashlib.
  - Sequential chunk send with progress callbacks.
  - Auto-sized chunks based on the device's advertised maxPayload.

Only useful when the firmware was compiled with `-DCONDUYT_OTA`. Check
``device.capabilities.ota_capable`` before starting; the device will NAK
``OTA_BEGIN`` with ``OTA_INVALID`` (0x0F) if OTA isn't built in.
"""

from __future__ import annotations

import hashlib
from typing import Awaitable, Callable, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from .device import ConduytDevice

ProgressCallback = Callable[[int, int], None]


class ConduytOTA:
    """Async helper that flashes a firmware image to a connected device."""

    def __init__(self, device: "ConduytDevice") -> None:
        self._device = device

    async def flash(
        self,
        firmware: bytes,
        *,
        chunk_size: Optional[int] = None,
        on_progress: Optional[ProgressCallback] = None,
        sha256: Optional[bytes] = None,
    ) -> None:
        """Flash a firmware image. Resolves once OTA_FINALIZE acks.

        The device typically reboots immediately after — the next round-trip
        will fail until you reconnect.

        Args:
            firmware: The firmware bytes to flash.
            chunk_size: Bytes per OTA_CHUNK. Defaults to (max_payload - 4).
            on_progress: Optional callback ``fn(sent, total)`` after each chunk.
            sha256: Optional pre-computed SHA-256 of `firmware`.
        """
        total = len(firmware)
        if total == 0:
            raise ValueError("OTA: firmware is empty")

        digest = sha256 if sha256 is not None else hashlib.sha256(firmware).digest()

        # Pick a chunk size. Reserve 4 bytes for the LE u32 offset header.
        caps = self._device.capabilities
        advertised = getattr(caps, "max_payload", None) if caps else None
        safe_max = max(64, advertised - 4) if advertised else 240
        size = chunk_size if chunk_size is not None else safe_max

        await self._device.ota_begin(total, digest)
        sent = 0
        while sent < total:
            end = min(sent + size, total)
            await self._device.ota_chunk(sent, firmware[sent:end])
            sent = end
            if on_progress:
                on_progress(sent, total)
        await self._device.ota_finalize()
